render/allocator.c

```diff
@@ -23,21 +24,35 @@ struct wlr_allocator *wlr_allocator_autocreate(struct wlr_backend *backend,
 	uint32_t backend_caps = backend_get_buffer_caps(backend);
 	uint32_t renderer_caps = renderer_get_render_buffer_caps(renderer);
 	int drm_fd = wlr_backend_get_drm_fd(backend);
+	int fd = -1;
+	if (drm_fd != -1) {
+		fd = fcntl(drm_fd, F_DUPFD_CLOEXEC, 0);
```

Comment:
Because the DRM dumb allocator is using `drmPrimeHandleToFD`, we can't `dup` the DRM FD here, or we risk double-closing GEM handles. See https://gitlab.freedesktop.org/mesa/drm/-/merge_requests/110 for more info.

Instead, we need to re-open the DRM FD when using the DRM dumb allocator. To do so, we can use `drmGetDeviceNameFromFd2` to get the path to the DRM node, use `wlr_backend_get_session` to get the `wlr_session` (fail if there's no session), and then use `wlr_session_open_file`. Maybe all of this mess can be isolated in a `attempt_drm_dumb_allocator` function or something.

Reply:
I took that bit of code from all of the backend, they previously did that operation before creating their allocators. What was preventing double closing GEM handles?

Reply:
Other backends and allocators don't use `drmPrimeHandleToFD`, so they can just `dup` the DRM FD. Since the DRM dumb allocator uses `drmPrimeHandleToFD`, it needs to re-open the DRM FD.